### PR TITLE
[client-py] add upload/download functionality

### LINKED ISSUE #4424: Add support for uploading and downloading to the Python clients

### PR DESCRIPTION
**Pull request by djmitche**
Github Bug/Issue: Fixes #4424

### CODE REVIEW

**Comment by djmitche** [comment]
https://github.com/mozilla/community-tc-config/pull/397 for the scope issue.

**Comment by djmitche** [comment]
This, and the other clients, will need some more configuration of the community deployment.  I'll get an issue filed.

**Comment by djmitche** [comment]
https://jira.mozilla.com/browse/SE-1570 for that

**Comment by djmitche** [comment]
Moved to https://bugzilla.mozilla.org/show_bug.cgi?id=1698094

File: clients/client-py/taskcluster/aio/upload.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                "contentType": contentType,
+                "objectData": encoded,
+            }
+
+        proposedUploadMethods['putUrl'] = {
+            "contentType": contentType,
+            "contentLength": contentLength,
+        }
+
+        uplodaResp = await objectService.createUpload(name, {
```
Comment (petemoore) [inline clients/client-py/taskcluster/aio/upload.py:86]:
```suggestion
        uploadResp = await objectService.createUpload(name, {
```

File: clients/client-py/taskcluster/aio/upload.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            "expires": expires,
+            "projectId": projectId,
+            "uploadId": uploadId,
+            "proposedUploadMethods": proposedUploadMethods,
+        })
+
+        async def tryUpload(retryFor):
+            try:
+                uploadMethod = uplodaResp["uploadMethod"]
```
Comment (petemoore) [inline clients/client-py/taskcluster/aio/upload.py:95]:
```suggestion
                uploadMethod = uploadResp["uploadMethod"]
```

File: clients/client-py/taskcluster/aio/upload.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+async def upload(*, projectId, name, contentType, contentLength, expires,
+                 readerFactory, maxRetries=5, objectService):
+    """
+    Upload the given data to the object service with the given metadata.
+    The `maxRetries` parameter has the same meaning as for service clients.
+    The `objectService` paraemter is an instance of the Object class,
+    configured with credentials for the upload.
+    """
+    async with aiohttp.ClientSession() as session:
+        uploadId = taskcluster.slugid.v4()
```
Comment (petemoore) [inline clients/client-py/taskcluster/aio/upload.py:68]: Maybe more future-proof to use `taskcluster.slugid.nice()` here? At the moment it probably isn't needed, but 121 bits entropy should still be plenty, and if a command line tool in future passes an `uploadId` as its first argument at some point in future, it has a little extra protection.

File: clients/client-py/taskcluster/aio/upload.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    chunk_size = 64 * 1024
+
+    async def reader_gen():
+        while True:
+            chunk = await reader.read(chunk_size)
+            if not chunk:
+                break
+            yield chunk
+
+    resp = await session.put(method['url'], headers=method['headers'], data=reader_gen())
```
Comment (petemoore) [inline clients/client-py/taskcluster/aio/upload.py:132]: If there is an error, does the caller have a way of logging the response body?

Comment (djmitche) [reply clients/client-py/taskcluster/aio/upload.py:132]: I believe the aiohttp exception raised in that case does contain the response, from which the body can be read.

Comment (petemoore) [reply clients/client-py/taskcluster/aio/upload.py:132]: oh cool!

File: clients/client-py/test/aio/test_client.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    })
+    result = await client.testAuthenticate({
+        'clientScopes': ['test:a'],
+        'requiredScopes': ['test:a'],
+    })
+    assert result == {'scopes': ['assume:anonymous', 'test:a'], 'clientId': 'tester'}
+
+
+@pytest.mark.asyncio
+async def test_async_works_with_temporary_credentials():
```
[the remark this answers is not in the document]
Comment (djmitche) [reply clients/client-py/test/aio/test_client.py:34]: This test isn't new..

[the remark this answers is not in the document]
Comment (djmitche) [reply clients/client-py/test/aio/test_client.py:34]: the file was renamed test_async.py -> test_client.py -- perhaps github didn't represent that?

File: clients/client-py/test/aio/test_upload_download.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            self.send_response(400)
+            self.end_headers()
+            self.wfile.write(b'uhoh')
+
+    with httptest.Server(Server) as ts:
+        objectService = FakeObject(ts)
+        with pytest.raises(aiohttp.ClientResponseError):
+            await download.downloadToBuf(
+                name="some/object",
+                objectService=objectService)
```
[the remark this answers is not in the document]
Comment (djmitche) [reply clients/client-py/test/aio/test_upload_download.py:71]: In fact, i got this wrong in the sync version!

File: clients/client-py/taskcluster/aio/upload.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    await upload(**kwargs, readerFactory=readerFactory)
+
+
+async def upload(*, projectId, name, contentType, contentLength, expires,
+                 readerFactory, maxRetries=5, objectService):
+    """
+    Upload the given data to the object service with the given metadata.
+    The `maxRetries` parameter has the same meaning as for service clients.
+    The `objectService` paraemter is an instance of the Object class,
```
Comment (jwhitlock) [inline clients/client-py/taskcluster/aio/upload.py:64]:
```suggestion
    The `objectService` parameter is an instance of the Object class,
```

File: clients/client-py/taskcluster/download.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        return file
+
+    return download(writerFactory=writerFactory, **kwargs)
+
+
+def download(*, name, maxRetries=5, objectService, writerFactory):
+    """
+    Download the named object from the object service, using a writer returned
+    from `writerFactory` to write the data.  The `maxRetries` parameter has
+    the same meaning as for service clients.  The `objectService` paraemter is
```
Comment (jwhitlock) [inline clients/client-py/taskcluster/download.py:62]:
```suggestion
    the same meaning as for service clients.  The `objectService` parameter is
```

File: clients/client-py/taskcluster/upload.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    upload(**kwargs, readerFactory=readerFactory)
+
+
+def upload(*, projectId, name, contentType, contentLength, expires,
+           readerFactory, maxRetries=5, objectService):
+    """
+    Upload the given data to the object service with the given metadata.
+    The `maxRetries` parameter has the same meaning as for service clients.
+    The `objectService` paraemter is an instance of the Object class,
```
Comment (jwhitlock) [inline clients/client-py/taskcluster/upload.py:62]:
```suggestion
    The `objectService` parameter is an instance of the Object class,
```

File: clients/client-py/test/aio/test_client.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+pytestmark = [
+    pytest.mark.skipif(os.environ.get("NO_TESTS_OVER_WIRE"), reason="Skipping tests over wire"),
+]
+
+
+@pytest.mark.asyncio
+async def test_async_works_with_permanent_credentials():
+    """we can call methods which require authentication with valid
+    permacreds"""
```
Comment (jwhitlock) [inline clients/client-py/test/aio/test_client.py:17]:
```suggestion
    """we can call methods which require authentication with valid permacreds"""
```

File: clients/client-py/test/test_upload_download.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+def test_putUrl_upload_fails(randbytes):
+    "When a putUrl upload's PUT fails with a 400, an exception is raised"
+    data = randbytes(10240)  # >8k to avoid using dataInline
+
+    class Server(httptest.Handler):
+        def do_PUT(self):
+            self.send_response(400)
+            self.end_headers()
+            self.wfile.write(b'uhoh')
```
Comment (jwhitlock) [inline clients/client-py/test/test_upload_download.py:135]: Should this test method also have an ``attempts`` counter with an assertion? I think 400 is the case that fails without retries.

**Review by jwhitlock — APPROVED** [review]
Thanks @djmitche, looks like good code (async and sync) plus good tests and documentation.

File: clients/client-py/test/test_upload_download.py (hunk 11 above)
Comment (djmitche) [reply clients/client-py/test/test_upload_download.py:135]: Indeed, and it appears this is also buggy!

Comment (djmitche) [reply clients/client-py/test/test_upload_download.py]: I addressed this (it was the same bug as pete noted in `taskcluster/download.py`).

**Review by petemoore — APPROVED** [review]

**Comment by djmitche** [comment]
Woo, thanks all!

**Pull request merged**